### PR TITLE
capi: Support u8 pixel format

### PR DESCRIPTION
**Pull request by lu-zero**
The hard part is done, mapping the capi::Packet can be made it work with a From impl

### CODE REVIEW

File: src/capi.rs [hunk 1]
```diff
@@ -78,6 +94,53 @@ pub struct Config {
...
+impl EncoderContext for rav1e::Context<u8> {
+  fn new_frame(&self) -> Frame {
+    rav1e::Context::<u8>::new_frame(self).into()
+  }
+  fn send_frame(&mut self, frame: Option<Frame>) -> Result<(), rav1e::EncoderStatus> {
+    if let Some(frame) = frame {
+      if let Frame::U8(f) = frame {
+        self.send_frame(f)
+      } else {
+        Err(rav1e::EncoderStatus::Failure)
```
Comment (dwbuiten) [inline src/capi.rs]: Won't this end up (once the other funcs are done) basically be two whole separate copied implementations?

That seems not ideal compared to having some sort of macro'd `match` inside each function, and having only *one* implementation...

Comment (lu-zero) [reply src/capi.rs]: Ideally we can go with the same pattern I applied to `rav1e_frame_fill_plane()`, I wanted to try the dyn Trait way as well and see what works better all in all.

``` rust
match (ctx, frame) {
    (Context::U8(c), Frame::U8(f)) => foo_internal(...),
    (Context::U16(c), Frame::U16(f)) => foo_internal(...),
    _ => SomeError
}
```

Probably is compact enough

Comment (dwbuiten) [reply src/capi.rs]: My preference is the latter, yeah - much more compact and only in one location.

File: src/capi.rs [hunk 2]
```diff
@@ -78,6 +94,53 @@ pub struct Config {
...
+trait EncoderContext {
+  fn new_frame(&self) -> Frame;
+  fn send_frame(&mut self, frame: Option<Frame>) -> Result<(), rav1e::EncoderStatus>;
+/*
+  fn receive_packet(&mut self) -> Packet;
+  fn container_sequence_header(&self) -> Vec<u8>;
+
+  fn twopass_bytes_needed(&self) -> usize;
+  fn twopass_in(&mut self, buf: &[u8]) -> Result<usize, EncoderStatus>;
+  fn twopass_out(&self) -> Vec<u8>;*/
```
Comment (kodabb) [inline src/capi.rs]: nit: extra comment cruft?

File: src/capi.rs [hunk 3]
```diff
@@ -435,13 +498,13 @@ pub unsafe extern "C" fn rav1e_twopass_in(ctx: *mut Context, buf: *mut u8, buf_s
 #[no_mangle]
 pub unsafe extern "C" fn rav1e_send_frame(ctx: *mut Context, frame: *const Frame) -> EncoderStatus {
     let frame = if frame.is_null() {
-        None
+      None
```
Comment (kodabb) [inline src/capi.rs]: nit: whitespaces

File: src/capi.rs (hunk 2 above)
Comment (lu-zero) [reply src/capi.rs]: It is what's left to be done :P

**Comment by lu-zero** [comment]
I added `receive_packet()` and I used the Trait to make so I the other functions can be still converted one by one and at the end we can swap the `Context<u16>` with the `enum` and everything should work properly.

**Comment by dwbuiten** [comment]
Tested and works, output is good, speedup is there.

File: src/capi.rs (hunk 3 above)
Comment (lu-zero) [reply src/capi.rs]: rustfmt will be run after that.

**Comment by lu-zero** [comment]
I squashed the whole thing and addressed the nit @kodabb, now we just have to wait for travis

**Pull request merged**